interfaces/input_parameters/ring.py

```diff
@@ -182,182 +130,348 @@ class Ring:
+        # Determining the momentum compaction orders defined in all sections
+        # The orders 1 and 2 are presently set by default to zeros if
+        # not defined in sections for the calculation of all orders of eta
+        self.alpha_orders = [
+            section.alpha_orders
+            for section in self.Section_list] + [[1], [2]]
 
-        # Momentum compaction, checks, and derived slippage factors
+        self.alpha_orders = np.unique(self.alpha_orders)
 
-        if not hasattr(alpha, '__iter__'):
-            alpha = (alpha, )
+        # Getting the momentum compaction programs in all sections
+        # and combining the programs (the missing programs are filled
+        # with zeros).
+        # The programs are reshaped to the size of the momentum program
+        for alpha_order in self.alpha_orders:
+            alpha_prog = []
+            alpha_name = 'alpha_%d' % (alpha_order)
+            for section in self.Section_list:
 
-        if isinstance(alpha, dict):
-            try:
-                if not all([k % 1 == 0 for k in alpha.keys()]):
-                    raise TypeError
-            except TypeError:
-                raise excpt.InputError("If alpha is dict all keys must be "
-                                       + "numeric and integer")
+                if hasattr(section, alpha_name):
+                    alpha_prog.append(getattr(section, alpha_name))
+                else:
+                    alpha_prog.append(ring_programs.momentum_compaction(
+                        0, order=alpha_order))
 
-            maxAlpha = np.max(tuple(alpha.keys())).astype(int)
-            alpha = [alpha.pop(i, 0) for i in range(maxAlpha+1)]
+            alpha_prog = ring_programs.momentum_compaction.combine_single_sections(
+                *alpha_prog)
 
-        if isinstance(alpha, dTypes._function):
-            alpha = (alpha,)
-
-        for i, a in enumerate(alpha):
-            if not isinstance(a, dTypes.momentum_compaction):
-                a = dTypes.momentum_compaction(a, order=i)
-
-            setattr(self, 'alpha_'+str(i), a.reshape(self.n_sections,
-                                                     self.cycle_time,
-                                                     self.use_turns))
-            setattr(self, 'eta_'+str(i), np.zeros([self.n_sections,
-                                                   len(self.use_turns)]))
-        self.alpha_order = i
-
-        for i in range(3 - self.alpha_order):
-            if not hasattr(self, f'alpha_{i}'):
-                setattr(self, 'alpha_'+str(i), np.zeros([self.n_sections,
-                                                         len(self.use_turns)]))
-                setattr(self, 'eta_'+str(i), np.zeros([self.n_sections,
-                                                       len(self.use_turns)]))
+            setattr(self, alpha_name, alpha_prog.reshape(
+                self.n_sections, self.cycle_time, self.use_turns))
```

Comment:
I think this can be simplified slightly:
```python
for alpha in self.alpha_orders:
    if alpha%1 == 0:
        alpha=int(alpha)
    else:
        raise excpt.BadAlphaOrderError("Only integers are allowed")
    alpha_prog = []
    for section in self.Section_list:
        try:
            alpha_prog.append(getattr(section, 'alpha_'+str(alpha)))
        except AttributeError:
            alpha_prog.append(ring_programs.momentum_compaction(
                        0, order=alpha_order))
        setattr(self, "alpha_"+str(alpha), ring_programs.momentum_compaction(*alpha_prog,
                                                              order=alpha).reshape(self.n_sections, self.cycle_time,
                                                                                                  self.use_turns))
```
or something to that effect.

Reply:
The check that alpha_orders are int is done in the RingSection object (but still could be consolidated there)

Will answer to the rest in the other PR